pkg/policy/repository.go

```diff
@@ -397,20 +424,86 @@ func (p *Repository) Add(r api.Rule) (uint64, map[uint16]struct{}, error) {
 // AddListLocked inserts a rule into the policy repository with the repository already locked
 // Expects that the entire rule list has already been sanitized.
 func (p *Repository) AddListLocked(rules api.Rules) (ruleSlice, uint64) {
-
-	newList := make(ruleSlice, len(rules))
-	for i := range rules {
+	newRules := make(ruleSlice, 0, len(rules))
+	for _, r := range rules {
 		newRule := &rule{
-			Rule:     *rules[i],
+			Rule:     *r,
+			key:      ruleKey{idx: p.nextID},
 			metadata: newRuleMetadata(),
 		}
-		newList[i] = newRule
+		newRules = append(newRules, newRule)
+		p.insert(newRule)
+		p.nextID++
+	}
+
+	return newRules, p.BumpRevision()
+}
+
+// ReplaceByResourceLocked replaces all rules that belong to a given resource with a
+// new set. The set of rules added and removed is returned, along with the new revision number.
+// Resource must not be empty
+func (p *Repository) ReplaceByResourceLocked(rules api.Rules, resource ipcacheTypes.ResourceID) (newRules ruleSlice, oldRules ruleSlice, revision uint64) {
+	if len(resource) == 0 {
+		// This should never ever be hit, as the caller should have already validated the resource.
+		// However, if it does happen, it means something very wrong has happened and we are at risk
+		// of removing all network policies. So, we must panic rather than risk disabling network security.
+		panic("may not replace API rules with an empty resource")
```

Comment:
Don't we have a more polite way to return an error or alert?

Reply:
We will intercept this earlier on; this is a last-gasp effort. Normally we could ignore the policy but in this case, I don't want to drop policies on the floor.